File: 2019/day25.go
package main

import (
	"bufio"
	"os"

	"github.com/lukechampine/advent/2019/intcode"
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day25_input.txt")
var prog = utils.ExtractInts(input)

func main() {
	// part 1
	stdin := bufio.NewReader(os.Stdin)
	m := intcode.New(prog)
	utils.Print(intcode.DecodeASCII(m.Run()))
	for !m.Halted {
		line, _, _ := stdin.ReadLine()
		input := string(line) + "\n"
		utils.Print(intcode.DecodeASCII(m.RunASCII(input)))
	}
}

File: 2019/day25.zig
const std = @import("std");
const utils = @import("utils.zig");

fn printASCII(ascii: []i64) void {
    for (ascii) |a| {
        if (a == '\n') {
            std.debug.warn("\n", .{}); // https://github.com/ziglang/zig/issues/4014
        } else {
            std.debug.warn("{c}", .{@intCast(u8, a)});
        }
    }
}

pub fn main() void {
    var prog = utils.loadProgram("day25_input.txt");
    var m = utils.Machine.init(prog);
    printASCII(m.run(@as([]const u8, "")) orelse unreachable);
    var buf: [100]u8 = undefined;
    while (true) {
        var line = std.io.readLineSlice(buf[0..]) catch unreachable;
        var input = utils.append(u8, line, '\n');
        printASCII(m.run(input) orelse unreachable);
    }
}
